src/main/java/org/opensearch/knn/index/mapper/ANNConfig.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.knn.index.mapper;
+
+import lombok.Getter;
+import org.opensearch.knn.index.engine.KNNMethodContext;
+
+import java.util.Optional;
+
+/**
+ * Class holds information about how the ANN indices are created. The design of this class ensures that we do not
+ * accidentally configure an index that has multiple ways it can be created. This class is immutable.
+ */
+public final class ANNConfig {
+
+    @Getter
+    private final ANNConfigType annConfigType;
+    private final KNNMethodContext knnMethodContext;
+    private final String modelId;
+    private final Integer dimension;
```

Comment:
> KNNMethodContext here is dense with layers in it. For instance model based context has KNNEngine inside of model metadata and in a layer above otherwise, this causes if else checks to get KNNEngine. Unwrapping KNNMethodContext and storing specific values will isolate different cases and logic associated with it while setting it rather than during get.

I prefer right now to keep it simple and encapsulate the 3 values from the mapping that can have null behavior or null-like behavior. I think this is something that can be circled back to in future.

> Is it worth considering to have specific configuration with hard types derived from ANNConfig. There is a case here where if modelId is present then dimension is -1. We can model it in a way where both fields are not in the same implementation. I am not sure how much this will help

Yes, I think I agree with this. Im going to switch ANNConfig to KNNMappingConfig and make it an interface.